faunadb-scala/src/main/scala/faunadb/values/CodecMacro.scala

```diff
@@ -0,0 +1,50 @@
+package faunadb.values
+
+import scala.reflect.macros.blackbox
+
+class CodecMacro(val c: blackbox.Context) {
+  import c.universe._
+
+  def customImpl[T: c.WeakTypeTag]: c.Tree = {
+    val tag = weakTypeOf[T]
+
+    val fields = getFields(tag)
+
+    val fieldsFragments = fields.map { field =>
+      fq"${varName(field)} <- value(${varName(field).toString}).to[${field._2}]"
+    }
+
+    val varNames = fields.map(varName)
+    val decodeFields = q"for (..$fieldsFragments) yield new $tag(..$varNames)"
+
+    val encodeFields = fields.map { field =>
+      q""" ${varName(field).toString} -> (value.${varName(field).toTermName}: Value) """
+    }
+
+    q""" new Codec[$tag] {
```

Comment:
As a general note, I've found this hard to understand because the building blocks precedes the actual macro expansion. IMO, it would be easier to understand this quote comes first and then we extract the building blocks as methods.